### ISSUE TITLE
Vertical Line Primitive Drifts on Chart Pan/Zoom

### ISSUE THREAD

**Issue by AVINASH0052** [tradingview/lightweight-charts#1935]
A vertical line added to the chart to mark a specific date does not remain fixed to that date. When I pans the chart horizontally or zooms in/out, the line moves along with the price data instead of staying anchored to its designated date on the time-axis.

The vertical line should remain static at the specific date it was drawn on. For example, if the line is drawn at `2025-10-20`, it should always align with that point on the time-axis, regardless of how the chart's visible range is changed by panning or zooming.

https://github.com/user-attachments/assets/90b83dcc-675a-47e7-a772-50be262b339a

I followed these resources for this implementation.

Live Example: https://tradingview.github.io/lightweight-charts/plugin-examples/plugins/vertical-line/example/

Example Source Code: https://github.com/tradingview/lightweight-charts/blob/master/plugin-examples/src/plugins/vertical-line/example/index.html

Official Documentation: https://tradingview.github.io/lightweight-charts/docs/next/plugins/intro (This page explains that primitives should be used to draw vertical lines that stick to the chart.)

```1. VerticalLine.js (The Primitive)
This file contains the core logic for the vertical line. It defines a custom object (a "primitive") that tells the charting library how to draw and position the line so it remains fixed to a specific date.

//This class implements a custom primitive for lightweight-charts to draw a vertical line.

export class VerticalLine {
    constructor(chart, series, time, options) {
        this._chart = chart;
        this._series = series;
        this._time = time;
        this._options = options;
        this._paneViews = [new VerticalLinePaneView(this)];
    }

    updateAllViews() {
        this._paneViews.forEach(view => view.update());
    }

    paneViews() {
        return this._paneViews;
    }

    detach() {
        if (this._series) {
            this._series.detachPrimitive(this);
        }
    }

    get time() { return this._time; }
    get options() { return this._options; }
    get chart() { return this._chart; }
}

// A dedicated renderer class for the vertical line.

class VerticalLineRenderer {
    constructor(source) {
        this._source = source;
    }

    draw(target) {
        target.useBitmapCoordinateSpace(scope => {
            const chart = this._source.chart;
            const time = this._source.time;
            const options = this._source.options;

            // Converts the date to a pixel coordinate on every frame, fixing the line's position.
            const x = chart.timeScale().timeToCoordinate(time);
            if (x === null) {
                return; // Don't draw if the time is not visible on the chart
            }

            const ctx = scope.context;
            const paneHeight = scope.bitmapSize.height;
            
            ctx.beginPath();
            ctx.moveTo(x, 0);
            ctx.lineTo(x, paneHeight);
            ctx.strokeStyle = options.color || 'white';
            ctx.lineWidth = options.width || 2;
            ctx.stroke();
        });
    }
}


// The PaneView manages the renderer for the vertical line.
 
class VerticalLinePaneView {
    constructor(source) {
        this._source = source;
        // Create a single, persistent renderer instance.
        this._renderer = new VerticalLineRenderer(this._source);
    }

    update() {}

    renderer() {
        return this._renderer;
    }
}

2. index.js (The Helper Function)
This file provides a clean helper function, addVerticalLine, that simplifies the process of creating and adding the VerticalLine primitive to the chart.

// From index.js

import { VerticalLine } from "./VerticalLine.js";
import { LineStyle } from "lightweight-charts";

// ... other functions

// Add a vertical line to a series using a primitive
function addVerticalLine(chart, series, time, options = {}) {
  const defaults = {
    color: 'yellow',
    width: 2,
    style: LineStyle.Solid
  };
  const lineOptions = { ...defaults, ...options };
  const verticalLinePrimitive = new VerticalLine(chart, series, time, lineOptions);
  series.attachPrimitive(verticalLinePrimitive);
  
  // Return the primitive itself so it can be detached later
  return verticalLinePrimitive;
}

// ... other exports
export { addVerticalLine };

3. PriceTragetChart.jsx (The Implementation)
This React component is where the addVerticalLine function is actually called. The code below automatically loops through all the projection data and calls addVerticalLine for each one to draw all the lines when the app loads.

// From PriceTragetChart.jsx

// ... other code

// 4. Automatically draw all projections when table data is set
useEffect(() => {
    if (!priceTargetTableData.length || !priceTargetChart || !lineSeries || !chartData?.length) return;

    const allPrimitives = [];

    // Loop through all available projections and draw them
    for (const data of priceTargetTableData) {
        const rawTargetDate = data.high_bound_target_date_end || data.low_bound_target_date_end;
        if (data.upload_date === rawTargetDate) {
            continue; 
        }

        // ... logic to find the start date for the projection line ...

        // This is where the vertical line is created for each projection
        if (rawTargetDate) {
            const targetDate = rawTargetDate.split('T')[0];
            const verticalLine = addVerticalLine(priceTargetChart, lineSeries, targetDate, { color: 'rgba(255, 255, 0, 0.4)', width: 1, style: 2 });
            allPrimitives.push({ detach: () => verticalLine.detach() });
        }

        // ... logic to draw the projection lines ...
    }
    setRemovablePrimitives(allPrimitives);
    
}, [priceTargetTableData, priceTargetChart, lineSeries, chartData]);


Data for the Main Line Graph
This data is used to plot the main blue line graph. It represents historical stock prices and would typically be in a file like public/tsla-data.json. The createLineSeries function plots the close value for each timestamp.

[
  {
    "timestamp": "2024-06-01",
    "open": 300.00,
    "high": 305.50,
    "low": 298.75,
    "close": 302.25
  },
  {
    "timestamp": "2024-06-02",
    "open": 302.25,
    "high": 310.00,
    "low": 301.50,
    "close": 308.70
  },
  {
    "timestamp": "2024-06-03",
    "open": 308.70,
    "high": 315.20,
    "low": 307.80,
    "close": 312.50
  },
  {
    "timestamp": "2024-06-04",
    "open": 312.50,
    "high": 314.00,
    "low": 305.10,
    "close": 306.90
  },
  {
    "timestamp": "2024-06-05",
    "open": 306.90,
    "high": 308.00,
    "low": 299.50,
    "close": 301.40
  }
]


Data for Vertical & Projection Lines
This data is used to plot the dashed projection lines and the vertical yellow lines. It represents future price targets and would typically be fetched from an API endpoint like /api/projections-data.

upload_date: The start date of the projection line.

high_bound_target_date_end / low_bound_target_date_end: The end date for both the projection line and the vertical line.

high_bound / low_bound: The target price at the end date.

{
  "summaries": [
    {
      "upload_date": "2024-06-02",
      "high_bound": 350.00,
      "high_bound_target_date_end": "2024-09-15T00:00:00.000Z",
      "low_bound": null,
      "low_bound_target_date_end": null
    },
    {
      "upload_date": "2024-06-04",
      "high_bound": null,
      "high_bound_target_date_end": null,
      "low_bound": 280.00,
      "low_bound_target_date_end": "2024-08-20T00:00:00.000Z"
    },
    {
      "upload_date": "2024-06-05",
      "high_bound": 330.00,
      "high_bound_target_date_end": "2024-10-01T00:00:00.000Z",
      "low_bound": null,
      "low_bound_target_date_end": null
    }
  ]
}

**Comment by SlicedSilver** [comment]
Most likely you are having a similar issue to this: https://github.com/tradingview/lightweight-charts/issues/1920#issuecomment-3059293994

Essentially you are are using `useBitmapCoordinateSpace` but not accounting for the device pixel ratio of the screen in your canvas code. (see here for more information: https://tradingview.github.io/lightweight-charts/docs/plugins/canvas-rendering-target#difference-between-bitmap-and-media)

It might be as simple to fix as changing this:
```
const x = chart.timeScale().timeToCoordinate(time);
```
to
```
const x = chart.timeScale().timeToCoordinate(time) * scope.horizontalPixelRatio;
```

Additionally there is an example plugin for a vertical line:
- Demo: https://tradingview.github.io/lightweight-charts/plugin-examples/plugins/vertical-line/example/
- Code: https://github.com/tradingview/lightweight-charts/tree/master/plugin-examples/src/plugins/vertical-line

**Comment by AVINASH0052** [comment]
Hey @SlicedSilver, thank you so much! That worked perfectly.

You were absolutely right about `devicePixelRatio`. Multiplying the coordinate by `scope.horizontalPixelRatio` fixed the issue immediately. I really appreciate you taking time to explain the problem and provided the solution!

<img width="1440" height="762" alt="Image" src="https://github.com/user-attachments/assets/4197b9aa-f639-4905-8d63-f4095895611d" />
